price/pricing/network_load_calculator.go

```diff
@@ -0,0 +1,242 @@
+		if err != nil {
+			log.Err(err).Msgf("could not update multiplier for %s", k)
+		}
+		newCache[k] = mul
+		log.Trace().Msgf("updated multiplier for %s to %v", k, mul)
+	}
+	nlmc.setMultiplierMapCache(newCache)
+	log.Debug().Msgf("country update complete")
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) updateMultiplierForCountryByLoad(isoCode ISO3166CountryCode) (float64, error) {
+	mul, err := nlmc.fetchMultiplierFromProm(isoCode)
+	if err != nil {
+		return mul, err
+	}
+
+	return mul, nlmc.storeMultiplier(isoCode, mul)
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) storeMultiplier(isoCode ISO3166CountryCode, multiplier float64) error {
+	conn, err := nlmc.db.Connection()
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	query := `INSERT INTO country_price_multipliers(
+		country_code, multiplier, updated_at)
+		VALUES ($1, $2, now())
+		ON CONFLICT (country_code) DO UPDATE
+			SET multiplier = $2,
+				updated_at = now();`
+	_, err = conn.Exec(ctx, query, isoCode.String(), multiplier)
+	return err
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) fetchMultiplierFromProm(isoCode ISO3166CountryCode) (float64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
+	defer cancel()
+	providers, err := nlmc.getActiveProvidersForCountry(ctx, isoCode)
+	if err != nil {
+		return 0, err
+	}
+
+	activeSessions, err := nlmc.getSessionsForCountry(ctx, isoCode)
+	if err != nil {
+		return 0, err
+	}
+
+	mul := nlmc.calculateMultiplier(providers, activeSessions)
+	return mul, nil
+}
+
+// calculateMultiplier calculates a coefficient for price multiplication.
+// It returns a value of 0.5 < x < 2.0.
+// If there are more active sessions than providers, we want to incentivise providers by increasing service price in that region.
+func (nlmc *NetworkLoadMultiplierCalculator) calculateMultiplier(providers, activeSessions int64) float64 {
+	if providers == 0 {
+		return 1
+	}
+
+	coeff := float64(activeSessions) / float64(providers)
+	if coeff < 0.5 {
+		return 0.5
+	}
+	if coeff > 2 {
+		return 2.0
+	}
+
+	return coeff
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) getActiveProvidersForCountry(ctx context.Context, isoCode ISO3166CountryCode) (int64, error) {
+	v, _, err := nlmc.promClient.Query(
+		ctx,
+		fmt.Sprintf("count(sum(proposal_event{country='%s'}[1h]) by (provider_id))", isoCode),
+		time.Now(),
+	)
+	if err != nil {
+		return 0, fmt.Errorf("could not fetch proposals for country %s", isoCode)
+	}
+
+	casted := v.(model.Vector)
+	if len(casted) == 0 {
+		return 0, nil
+	}
+
+	return int64(casted[0].Value), nil
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) getSessionsForCountry(ctx context.Context, isoCode ISO3166CountryCode) (int64, error) {
+	v, _, err := nlmc.promClient.Query(
+		ctx,
+		fmt.Sprintf("count(count(session_data{provider_country='%s'}[1h]) by (session_id))", isoCode),
```

Comment:
This could be slow, I think it is better to aggregate it to the new metric.

Reply:
only calling this once half an hour, and responses seem rather fast. So i think it's OK.